wolfcrypt/src/port/arm/armv8-32-poly1305.c

```diff
@@ -0,0 +1,1193 @@
+  armv8_32_poly1305_blocks        This routine implements the 
+                        poly1305 algorithm.  The processing
+                        of the 16 byte pieces of the 
+                        message is implemented in assembly.
+                        Care was taken in implementing the
+                        assembly to optimize speed by 
+                        minimizing memory references as
+                        well as minimizing instructions.
+                        The poly1305 algorithm is implemented
+                        by using multiple 26 bit values 
+                        to represent the larger 16 byte and 
+                        17 byte numbers.  Breaking the 
+                        larger numbers down into 26 bit 
+                        pieces means that math operations will
+                        not overflow the 32 bit registers.
+                        For more information reference the article 
+                        NEON Cryto by Daniel Berstein and Peter
+                        Schwabe.
+
+  ctx                   poly1305 structure containing the r and
+                        h values broke into 26 bit pieces.
+                        The result of this routine is returned
+                        int the h array of 26 bit values.
+  msgData               Message data to run the algorithm 
+                        against.
+  keyData               256 bit key that is broken down 
+                        into the r and s subcomponents.
+**********************************************************/
+/* variable assembly code creates on stack */
+
+#define BYTES_SP_OFF              20
+#define HIBIT_LOCAL_SP_OFF        24
+
+void armv8_32_poly1305_blocks(Poly1305* ctx, 
+                              const unsigned char *msgData,
+                              size_t msgDataLen);
+
+/* WOLFSSL_ARMASM_NO_NEON is used in armv8-32-sha512-asm.S */
+#ifdef WOLFSSL_ARMASM_NO_NEON2
+void armv8_32_poly1305_blocks(Poly1305* ctx, 
+                              const unsigned char *msgData,
+                              size_t msgDataLen)
+{
+        /* these variables will be referenced via r7 in the assembly */
+        unsigned r[5], *r_ptrLocal, h[5], *h_ptrLocal;
+        unsigned msgDataOnStack = (unsigned)msgData;
+        unsigned msgDataLenOnStack = msgDataLen;
+        unsigned hibitOnStack = (ctx->finished) ? 0 : ((unsigned) 1 << 24);
+
+        memcpy(r, ctx->r, 5 * sizeof(unsigned));
+        memcpy(h, ctx->h, 5 * sizeof(unsigned));
+
+        /* 32 bit pointer, this is necessary to pass the pointer value
+           into the assembly code */
+        r_ptrLocal = (unsigned *) r;
+        h_ptrLocal = (unsigned *) h;
+
+#ifdef POLY1305_VERBOSE
+        int i;
+        for (i=0;i<5;++i) printf("h%d %07X ", i, h[i]);
+        printf("\n");
+
+        printf("armv8_32_poly1305_blocks() data len %d\n", msgDataLen);
+#endif
+
+        __asm__ __volatile__ (
```

Comment:
Can you show me where the "neon" instructions are used? This is not much faster than the C version.

Reply:
I agree.

I talked to Sean about the algorithm.  I am disappointed that the speed up is only 12.8 %.

VMULL and VLMAL  allow two 32 bit values to be multiplied to create one 64 bit value.  I did not get a vector version of the instruction to work and I did try encoding a few instructions by hand.

The only speed up potential in the NEON version submitted comes from the fact that the memory references in the non-NEON version are replaced by use of NEON registers.....